### PR TITLE
Cheetah3D -> BabylonJS Exporter Update

### PR DESCRIPTION
**Pull request by peterkingsbury**
Main changes:
- all of everything is ES6
- classes for Babylon.js entities
- fixed export hierarchy bug in meshes
- added specular, emissive, bump, reflection and opacity map support 
- rearranged file save to occur before any heavy work is done
- added a hierarchical scene text-output to visualize entity hierarchy
- documented lights, replaced one kind of point light with a hemispherical light
- added getFunctions() and getParameters() for debugging Cheetah3D objects

### CODE REVIEW

**Comment by voidtuxic** [comment]
hey! Thanks for your work, code looks great now :) I'll get my Cheetah3D setup back on to test it, check the bug you pinged me about on twitter!

File: Babylon file export.js [hunk 1]
```diff
@@ -1,43 +1,45 @@
...
+    this.id = obj.getParameter('name');
+    let vec3 = obj.getParameter('position');
     this.position = [vec3.x, vec3.y, vec3.z];
-    vec3 = obj.getParameter("rotation");
-    this.rotation = [vec3.y* Math.PI/ 180, -vec3.x* Math.PI/ 180, -vec3.z* Math.PI/ 180];
-    vec3 = obj.getParameter("scale");
+    vec3 = obj.getParameter('rotation');
+    this.rotation = [vec3.y * Math.PI / 180, -vec3.x * Math.PI / 180, -vec3.z * Math.PI / 180];
+
+    // this.rotation = [vec3.x, vec3.y, vec3.z];
```
Comment (voidtuxic) [inline Babylon file export.js]: comment cleanup

File: Babylon file export.js [hunk 2]
```diff
@@ -50,308 +52,528 @@ var BabylonMesh = function(obj)
...
+      disableDepthWrite: MATERIAL_DISABLE_DEPTH_WRITE,
+      ambient:           new BabylonColor3(0, 0, 0).export(),
+      diffuse:           new BabylonColor3(diffuse.x, diffuse.y, diffuse.z).export(),
+      specular:          new BabylonColor3(specular.x, specular.y, specular.z).export(),
+      specularPower:     material.shininess(),
+      emissive:          new BabylonColor3(emission.x, emission.y, emission.z).export(),
+      alpha:             1.0,
+      backFaceCulling:   true,
+      wireframe:         false,
+      /* diffuseTexture:                   new BabylonTexture(),
```
Comment (voidtuxic) [inline Babylon file export.js]: comment cleanup (unless this is temporary?)

File: Babylon file export.js [hunk 3]
```diff
@@ -50,308 +52,528 @@ var BabylonMesh = function(obj)
...
+        let camera = new BabylonCamera(child);
+        cameras.push(camera);
+      } break;
+
+      default: {
+        let mesh = new BabylonMesh(child);
+        if (parentId) {
+          mesh.parentId = parentId;
+        }
+        // parentId = mesh.id;
```
Comment (voidtuxic) [inline Babylon file export.js]: comment cleanup

File: Babylon file export.js [hunk 4]
```diff
@@ -50,308 +52,528 @@ var BabylonMesh = function(obj)
...
+ * @param {Object} doc The Cheetah3D document object.
+ */
+function main(doc) {
+  clearConsole();
+  let path = OS.runSavePanel('babylon');
+  if (path == null) {
+    return;
+  }
+
+//  let path = `/Users/peter/Development/galactikore/public/src/media/scenes/testScene.babylon`;
```
Comment (voidtuxic) [inline Babylon file export.js]: comment cleanup

**Review by voidtuxic — COMMENTED** [review]
just being picky on your wip comments, now that I'm seeing good es7 on this project :) other than that, I haven't checked in detail the code but it's looking really good! It works great with the latest version of Cheetah3D, I'll admit I didn't take the time to test previous versions. Feel free to add your name to the readme for credit, and if you want to clean those comments :) in any case, tell me when you're good for merging, I'll go ahead! thanks!

File: Babylon file export.js (hunk 2 above)
Comment (peterkingsbury) [reply Babylon file export.js]: I'll remove it; it's mostly just commented code I derived from the Babylon docs. It was helpful to me to see all exposed attributes.

**Comment by peterkingsbury** [comment]
Think I got everything... should be ready for merge.

**Comment by voidtuxic** [comment]
seems great! :shipit:

**Pull request merged**